adafruit_azureiot/iot_mqtt.py

```diff
@@ -0,0 +1,453 @@
+            signature = signature[:-1]
+        token = "SharedAccessSignature sr={}&sig={}&se={}".format(uri, signature, token_expiry)
+        return token
+
+    # Workaround for https://github.com/adafruit/Adafruit_CircuitPython_MiniMQTT/issues/25
+    def _try_create_mqtt_client(self, hostname: str) -> None:
+        minimqtt.set_socket(socket, self._wifi_manager.esp)
+
+        self._mqtts = MQTT(
+            broker=hostname,
+            username=self._username,
+            password=self._passwd,
+            port=8883,
+            keep_alive=120,
+            is_ssl=True,
+            client_id=self._device_id,
+            log=True,
+        )
+
+        self._mqtts.logger.setLevel(self._logger.getEffectiveLevel())
+
+        # set actions to take throughout connection lifecycle
+        self._mqtts.on_connect = self._on_connect
+        self._mqtts.on_message = self._on_message
+        self._mqtts.on_log = self._on_log
+        self._mqtts.on_publish = self._on_publish
+        self._mqtts.on_disconnect = self._on_disconnect
+
+        # initiate the connection using the adafruit_minimqtt library
+        self._mqtts.last_will()
+        self._mqtts.connect()
+
+    def _create_mqtt_client(self) -> None:
+        try:
+            self._try_create_mqtt_client(self._hostname)
+        except ValueError:
+            # Workaround for https://github.com/adafruit/Adafruit_CircuitPython_MiniMQTT/issues/25
+            self._try_create_mqtt_client("https://" + self._hostname)
+
+    # pylint: disable=C0103, W0613
+    def _on_connect(self, client, userdata, _, rc) -> None:
+        self._logger.info("- iot_mqtt :: _on_connect :: rc = " + str(rc) + ", userdata = " + str(userdata))
+        if rc == 0:
+            self._mqtt_connected = True
+        self._auth_response_received = True
+        self._callback.connection_status_change(True)
+
+    # pylint: disable=C0103, W0613
+    def _on_log(self, client, userdata, level, buf) -> None:
+        self._logger.info("mqtt-log : " + buf)
+        if level <= 8:
+            self._logger.error("mqtt-log : " + buf)
+
+    def _on_disconnect(self, client, userdata, rc) -> None:
+        self._logger.info("- iot_mqtt :: _on_disconnect :: rc = " + str(rc))
+        self._auth_response_received = True
+
+        if rc == 5:
+            self._logger.error("on(disconnect) : Not authorized")
+            self.disconnect()
+
+        if rc == 1:
+            self._mqtt_connected = False
+
+        if rc != 5:
+            self._callback.connection_status_change(False)
+
+    def _on_publish(self, client, data, topic, msg_id) -> None:
+        self._logger.info("- iot_mqtt :: _on_publish :: " + str(data) + " on topic " + str(topic))
+
+    # pylint: disable=W0703
+    def _handle_device_twin_update(self, msg: str, topic: str) -> None:
+        self._logger.debug("- iot_mqtt :: _echo_desired :: " + topic)
+        twin = None
+        desired = None
+
+        print(msg)
```

Comment:
may want to log this instead

Reply:
Good spot - they were left in during debugging as some of the logging cause the logger to barf with format errors. Now removed.

Reply:
np! We generally never print from libraries, only user-code